File: src/core/api/device.ts
/**
 * Device API - Device management
 */
import { invoke } from '@tauri-apps/api/core';
import type { ApiResponse, DeviceInfo } from './types';

/**
 * Get list of available devices
 */
export async function getDevices(): Promise<ApiResponse<DeviceInfo[]>> {
  try {
    return await invoke<ApiResponse<DeviceInfo[]>>('device_list');
  } catch (error) {
    console.error('[API] getDevices error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Get device utilization stats
 */
export async function getDeviceStats(deviceId: number): Promise<ApiResponse<{ gpu_util: number; memory_util: number; temperature: number }>> {
  try {
    return await invoke<ApiResponse<{ gpu_util: number; memory_util: number; temperature: number }>>('device_stats', { deviceId });
  } catch (error) {
    console.error('[API] getDeviceStats error:', error);
    return { success: false, error: String(error) };
  }
}

/**
 * Set default training device
 */
export async function setDefaultDevice(deviceId: number): Promise<ApiResponse<void>> {
  try {
    return await invoke<ApiResponse<void>>('device_set_default', { deviceId });
  } catch (error) {
    console.error('[API] setDefaultDevice error:', error);
    return { success: false, error: String(error) };
  }
}

File: src-tauri/src/modules/yolo/commands/device.rs
use crate::modules::yolo::services::device::{self, DeviceInfo, DeviceStats};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct CommandResponse<T> {
    pub success: bool,
    pub data: Option<T>,
    pub error: Option<String>,
}

impl<T> CommandResponse<T> {
    pub fn ok(data: T) -> Self {
        Self {
            success: true,
            data: Some(data),
            error: None,
        }
    }
    pub fn err(msg: String) -> Self {
        Self {
            success: false,
            data: None,
            error: Some(msg),
        }
    }
}

/// List available compute devices (GPU/CPU)
#[tauri::command]
pub fn device_list() -> CommandResponse<Vec<DeviceInfo>> {
    let devices = device::list_devices();
    CommandResponse::ok(devices)
}

/// Get device utilization statistics
#[tauri::command]
pub fn device_stats(device_id: i32) -> CommandResponse<DeviceStats> {
    // Only GPU devices have detailed stats
    if device_id < 0 {
        return CommandResponse::err("Invalid device ID".to_string());
    }

    match device::get_gpu_stats(device_id) {
        Some(stats) => CommandResponse::ok(stats),
        None => CommandResponse::ok(DeviceStats {
            gpu_util: 0.0,
            memory_util: 0.0,
            temperature: 0.0,
        }),
    }
}

/// Set the default training device (persisted in settings)
#[tauri::command]
pub fn device_set_default(device_id: i32) -> CommandResponse<()> {
    // For now, just acknowledge - actual persistence is handled by settings system
    CommandResponse::ok(())
}
